test/val/val_decoration_test.cpp

```diff
@@ -0,0 +1,160 @@
+    ; Since %1 is used first in Decoration, it gets id 1.
+)";
+  const uint32_t id = 1;
+  CompileSuccessfully(spirv);
+  EXPECT_EQ(SPV_SUCCESS, ValidateAndRetrieveValidationState());
+  // Must have 2 decorations.
+  EXPECT_THAT(vstate_->id_decorations(id),
+              Eq(vector<Decoration>{
+                  Decoration(SpvDecorationArrayStride, vector<uint32_t>{4}),
+                  Decoration(SpvDecorationUniform)}));
+}
+
+TEST_F(ValidateDecorations, ValidateOpMemberDecorateRegistration) {
+  string spirv = R"(
+    OpCapability Shader
+    OpCapability Linkage
+    OpMemoryModel Logical GLSL450
+    OpDecorate %_arr_double_uint_6 ArrayStride 4
+    OpMemberDecorate %_struct_115 2 NonReadable
+    OpMemberDecorate %_struct_115 2 Offset 2
+    OpDecorate %_struct_115 BufferBlock
+    %float = OpTypeFloat 32
+    %uint = OpTypeInt 32 0
+    %uint_6 = OpConstant %uint 6
+    %_arr_double_uint_6 = OpTypeArray %float %uint_6
+    %_struct_115 = OpTypeStruct %float %float %_arr_double_uint_6
+)";
+  CompileSuccessfully(spirv);
+  EXPECT_EQ(SPV_SUCCESS, ValidateAndRetrieveValidationState());
+
+  // The array must have 1 decoration.
+  const uint32_t arr_id = 1;
+  EXPECT_THAT(vstate_->id_decorations(arr_id),
+              Eq(vector<Decoration>{
+                  Decoration(SpvDecorationArrayStride, vector<uint32_t>{4})}));
+
+  // The struct must have 3 decorations.
+  const uint32_t struct_id = 2;
+  EXPECT_THAT(vstate_->id_decorations(struct_id),
+              Eq(vector<Decoration>{
+                  Decoration(SpvDecorationNonReadable, vector<uint32_t>{}, 2),
+                  Decoration(SpvDecorationOffset, vector<uint32_t>{2}, 2),
+                  Decoration(SpvDecorationBufferBlock)}));
+}
+
+TEST_F(ValidateDecorations, ValidateGroupDecorateRegistration) {
+  string spirv = R"(
+               OpCapability Shader
+               OpCapability Linkage
+               OpMemoryModel Logical GLSL450
+               OpDecorate %1 DescriptorSet 0
+               OpDecorate %1 NonWritable
+               OpDecorate %1 Restrict
+          %1 = OpDecorationGroup
+               OpGroupDecorate %1 %2 %3
+               OpGroupDecorate %1 %4
+  %float = OpTypeFloat 32
+%_runtimearr_float = OpTypeRuntimeArray %float
+  %_struct_9 = OpTypeStruct %_runtimearr_float
+%_ptr_Uniform__struct_9 = OpTypePointer Uniform %_struct_9
+         %2 = OpVariable %_ptr_Uniform__struct_9 Uniform
+ %_struct_10 = OpTypeStruct %_runtimearr_float
+%_ptr_Uniform__struct_10 = OpTypePointer Uniform %_struct_10
+         %3 = OpVariable %_ptr_Uniform__struct_10 Uniform
+ %_struct_11 = OpTypeStruct %_runtimearr_float
+%_ptr_Uniform__struct_11 = OpTypePointer Uniform %_struct_11
+         %4 = OpVariable %_ptr_Uniform__struct_11 Uniform
+  )";
+  CompileSuccessfully(spirv);
+  EXPECT_EQ(SPV_SUCCESS, ValidateAndRetrieveValidationState());
+
+  // Decoration group has 3 decorations.
+  auto check_decorations_for_id = [this](const uint32_t obj_id) {
```

Comment:
Putting the EXPECT_THAT inside a function will obscure the location of the failure if/when failure occurs.

Alternately ,to save typing you could assign the vector<Decoration> value to a constant value, and then do three EXPECT_THATs.